grpc/src/main/java/io/stargate/grpc/idempotency/IdempotencyAnalyzer.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright The Stargate Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.stargate.grpc.idempotency;
+
+import io.stargate.proto.QueryOuterClass;
+import java.util.Arrays;
+import java.util.List;
+import java.util.regex.Pattern;
+
+public class IdempotencyAnalyzer {
+  private static final List<String> NON_IDEMPOTENT_FUNCTIONS = Arrays.asList("now()", "uuid()");
+  private static final Pattern DELETE_FROM_MAP_PATTERN =
+      Pattern.compile("[a-zA-Z0-9 ] \\["); // any table name followed by white space and '['
+
+  public static boolean isIdempotent(List<QueryOuterClass.BatchQuery> queries) {
+    for (QueryOuterClass.BatchQuery query : queries) {
+      boolean idempotent = isIdempotent(query.getCql());
+      // if ANY query within a batch is non-idempotent, then the whole batch is non-idempotent
+      if (!idempotent) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  public static boolean isIdempotent(String query) {
```

Comment:
As you can see, the implementation is convoluted and complex. I am not sure if we will be able to catch all cases. At this point, I think that we should consider the same option as java-driver: that is, the end-user providing the `is_idempotent` when issuing a query with a default = `false` same as the driver:
```
By default, all DataStax Drivers consider queries to be non-idempotent. It is the user's responsibility to mark queries as idempotent to leverage features such as retry and speculative execution.
```
We may override/set it to `true` for all `SELECT` queries.
@olim7t, what was the main reason why the java-driver didn't decide to infer it from a query? (I think that I know the answer - the complexity but I wanted to double-check)

Reply:
> what was the main reason why the java-driver didn't decide to infer it from a query? (I think that I know the answer - the complexity but I wanted to double-check)

Yes, the complexity, and unfortunately we're running into the same kind of issue here. I'm afraid simple text searches are not going to be robust enough. Here are two queries that are not idempotent, yet they enter the first `if` block below:
```
INSERT INTO tbl(id, v) VALUES (uuid(), 'SELECT');
INSERT INTO a_table_named_select (id) VALUES (uuid());
```

To reliably analyze a CQL query string we need to parse it, but doing that in high-level modules like gRPC introduces two issues:
* duplication of the parsing logic that already exists in the Cassandra code
* the CQL grammar can change depending on the backend

I agree that we should do this further down the stack, but even our persistence modules are not deep enough: `Connection.prepare` is still dealing with a raw string, after that it's handling it to the regular (i.e. not overridden in Stargate) server code.

So we'd need to patch Cassandra itself, there is already a ticket for that: [CASSANDRA-10813](https://issues.apache.org/jira/browse/CASSANDRA-10813).

Reply:
PS- maybe we can do it in our persistence modules. That would take care of the second issue (CQL grammar specific to backend).
The double parsing would still happen, but maybe it's not that much of an overhead.

Reply:
Taking all of that into consideration, I think that at this stage of gRPC API evolution we should go with the simplest approach and let the caller provide this info (same as java-driver, here is the PR:
https://github.com/stargate/stargate/pull/1197)
please let me know wdyt

Reply:
Makes sense to me, given all the complexities. Worth revisiting if and when lower layer(s) can provide necessary information, and not have to work on string/regexp heuristics that may or may not work.